### PR TITLE
[DeepTest] CUBIC Congestion Control Test Suite Enhancement

### PR DESCRIPTION
**Pull request by saikat107**
## Summary

Rewrote `src/core/unittest/CubicTest.cpp` from 17 standalone tests (628 lines) to 50 fixture-based tests + 16 parametric `CubeRootTest` cases (1754 lines).

## What changed

- **Test fixture** with shared helpers (`InitializeDefaultWithRtt`, `EnterCongestionAvoidance`, `EnterHyStartActive`, `MakeAckEvent`/`MakeLossEvent`) and named RTT constants for HyStart thresholds.
- **16 parametric `CubeRootTest` cases** covering boundaries, perfect/non-perfect cubes, and large inputs.
- **HyStart++ state transition table** in-file, mapping every transition to its test.

## Test areas

| Area | Tests | What's covered |
|------|-------|----------------|
| Core CUBIC lifecycle | 14 | Initialization, reset, data sent/invalidated/acked, loss, ECN, recovery entry/exit, fast convergence, spurious congestion, zero-byte ACK, statistics, app-limited API |
| Congestion avoidance & AIMD | 7 | CUBIC formula vs AIMD selection, AIMD accumulator (half-rate, full-rate, overflow), sequential linear convergence, Reno-friendly region, CUBIC overflow to BytesInFlightMax |
| Pacing | 3 | Slow start estimation (2×CW), congestion avoidance estimation (1.25×CW), LastSendAllowance carryover |
| HyStart++ state machine | 16 | All 6 state transitions (NOT_STARTED→ACTIVE, NOT_STARTED→DONE, ACTIVE→DONE×2, ACTIVE→NOT_STARTED, DONE absorbing), delay increase/decrease detection, Eta calculation, conservative round counting, disabled mode, state persistence across recovery/spurious events, growth divisor invariant |
| Edge cases & boundaries | 8 | Minimum window floor, 2×BytesInFlightMax clamp, bounded growth clamp (1.5×CW), time gap overflow, persistent congestion window reset, double-loss guard, recovery stay-on-old-ACK, ECN spurious rollback asymmetry |
| Send allowance & control | 2 | CanSend scenarios (7 states), GetSendAllowance scenarios (5 states) |

## Coupling classification

| Category | Tests |
|----------|-------|
| Spec-conformance (RFC 8312, RFC 9002, HyStart++) | 23 |
| Implementation-specific, API-level assertions | 7 |
| Implementation-specific, internal struct assertions | 20 |

## Testing

```
[==========] 66 tests from 2 test suites ran.
[  PASSED  ] 66 tests.
```

### CODE REVIEW

File: src/core/congestion_control.h [hunk 1]
```diff
@@ -411,4 +411,4 @@ QuicCongestionControlSetAppLimited(
     )
 {
     Cc->QuicCongestionControlSetAppLimited(Cc);
-}
+}
```
Comment (guhetier) [inline src/core/congestion_control.h]: nit: we try to keep the end of file carriage return

File: src/core/unittest/CubicTest.cpp [hunk 2]
```diff
@@ -394,10 +419,19 @@ TEST(CubicTest, ResetScenarios)
...
     // Scenario 1: Partial reset (FullReset=FALSE) - preserves BytesInFlight
-    Cubic->BytesInFlight = 5000;
-    Cubic->SlowStartThreshold = 10000;
-    Cubic->IsInRecovery = TRUE;
-    Cubic->HasHadCongestionEvent = TRUE;
+    // First, send some data and trigger a congestion event to set internal flags
+    Connection.CongestionControl.QuicCongestionControlOnDataSent(&Connection.CongestionControl, 5000);
+
+    // Trigger congestion event via loss
+    QUIC_LOSS_EVENT LossEvent;
```
Comment (guhetier) [inline src/core/unittest/CubicTest.cpp]: Default initialization in C++ is better done with `QUIC_LOSS_EVENT LossEvent{};`.
In C, `QUIC_LOSS_EVENT LossEvent = {0};` is also generally enough.

File: src/core/unittest/CubicTest.cpp [hunk 3]
```diff
@@ -505,9 +550,10 @@ TEST(CubicTest, OnDataAcknowledged_BasicAck)
     Settings.SendIdleTimeoutMs = 1000;
 
     InitializeMockConnection(Connection, 1280);
+    Connection.Settings.HyStartEnabled = TRUE;  // Must set on Connection for runtime checks
```
Comment (guhetier) [inline src/core/unittest/CubicTest.cpp]: I am not sure I understand why we need HyStart.

File: src/core/unittest/CubicTest.cpp [hunk 4]
```diff
@@ -505,9 +550,10 @@ TEST(CubicTest, OnDataAcknowledged_BasicAck)
...
     Connection.Paths[0].GotFirstRttSample = TRUE;
     Connection.Paths[0].SmoothedRtt = 50000; // 50ms in microseconds
-
+    Connection.Settings.NetStatsEventEnabled = TRUE; // Enable logging
```
Comment (guhetier) [inline src/core/unittest/CubicTest.cpp]: Enabling this is probably good for coverage, but it seems there should either be a callback set to validate the data in the event, or a comment explaining this intentionally tests the case where no callback is present.

File: src/core/unittest/CubicTest.cpp [hunk 5]
```diff
@@ -759,13 +817,2265 @@ TEST(CubicTest, HyStart_StateTransitions)
...
+        &Connection.CongestionControl,
+        &PersistentLoss);
+
+    // After persistent congestion:
+    // - Window is reset to 2 * MTU (approximately 2560, but may vary)
+    // - Threshold is set to old_window * 0.7
+    uint32_t WindowAfterPC = Cubic->CongestionWindow;
+    uint32_t ThresholdAfterPC = Cubic->SlowStartThreshold;
+
+    ASSERT_LT(WindowAfterPC, 3000u); // Should be around 2*MTU
```
Comment (guhetier) [inline src/core/unittest/CubicTest.cpp]: This should be an equality. The test can control the MTU exactly.
A comparison brings little validation, if we were to make a mistake, it would be a coin-flip whether it does in the right direction or not to cause a test failure.

File: src/core/unittest/CubicTest.cpp [hunk 6]
```diff
@@ -759,13 +817,2265 @@ TEST(CubicTest, HyStart_StateTransitions)
...
+    AckEvent.IsLargestAckedPacketAppLimited = FALSE;
+    AckEvent.AdjustedAckTime = AckEvent.TimeNow;
+    AckEvent.AckedPackets = NULL;
+
+    Connection.CongestionControl.QuicCongestionControlOnDataAcknowledged(
+        &Connection.CongestionControl,
+        &AckEvent);
+
+    // Window should be valid (overflow was handled)
+    ASSERT_GT(Cubic->CongestionWindow, 0u);
```
Comment (guhetier) [inline src/core/unittest/CubicTest.cpp]: What is this validating? `CongestionWindow` is a uint32_t, so all this ensures is that the value is neither `0` nor `UINT32_MAX`.

The test seems to assume `CongestionWindow` is a signed value, like an int64_t maybe.

File: src/core/unittest/CubicTest.cpp [hunk 7]
```diff
@@ -759,13 +817,2265 @@ TEST(CubicTest, HyStart_StateTransitions)
...
+        if (round < QUIC_CONSERVATIVE_SLOW_START_DEFAULT_ROUNDS - 1) {
+            // Still in HYSTART_ACTIVE for first n-1 rounds
+            ASSERT_EQ(Cubic->HyStartState, HYSTART_ACTIVE);
+        } else {
+            // for the rest of the rounds, state should transition
+            // to HYSTART_DONE and stay there
+            ASSERT_EQ(Cubic->HyStartState, HYSTART_DONE);
+            ASSERT_EQ(Cubic->ConservativeSlowStartRounds, 0u);
+            // Verify SlowStartThreshold is set to current congestion window
+            ASSERT_LT(Cubic->SlowStartThreshold, UINT32_MAX);
```
Comment (guhetier) [inline src/core/unittest/CubicTest.cpp]: `SlowStartThreshold` is a uint32, this is not validating much.

**Comment by guhetier** [comment]
A general comment is that while the tests do provide coverage (which is good), the validation itself seems pretty weak in term of catching logic issues or regressions.
And in terms of finding undefined behavior, memory issues, etc... this probably doesn't beat fuzzing.

File: src/core/unittest/CubicTest.cpp (hunk 5 above)
Comment (saikat107) [reply src/core/unittest/CubicTest.cpp]: Resolved

File: src/core/unittest/CubicTest.cpp (hunk 7 above)
Comment (saikat107) [reply src/core/unittest/CubicTest.cpp]: Resolved with `ASSERT_EQ`

File: src/core/unittest/CubicTest.cpp (hunk 4 above)
Comment (saikat107) [reply src/core/unittest/CubicTest.cpp]: Not sure how to set the callbacks. Agent was not able to resolve that either. Added the comment -- please suggest if (and how) we should add callbacks.

File: src/core/unittest/CubicTest.cpp (hunk 3 above)
Comment (saikat107) [reply src/core/unittest/CubicTest.cpp]: You are right -- this is not necessary for this test case.

**Comment by saikat107** [comment]
Thanks for the review @guhetier -- these do not only make the current test cases better, but also the DeepTest agent better.

**Comment by saikat107** [comment]
@guhetier this PR is ready for review.

File: src/core/unittest/CubicTest.cpp [hunk 8]
```diff
@@ -864,7 +895,7 @@ TEST(CubicTest, Recovery_ExitOnNewAck)
     Connection.CongestionControl.QuicCongestionControlOnDataSent(&Connection.CongestionControl, 5000);
     Connection.Send.NextPacketNumber = 10; // Set packet number before loss
 
-    QUIC_LOSS_EVENT LossEvent;
+    QUIC_LOSS_EVENT LossEvent{};
     CxPlatZeroMemory(&LossEvent, sizeof(LossEvent));
```
Comment (guhetier) [inline src/core/unittest/CubicTest.cpp]: nit: The previous comment about zero-initializing using `{}` was addressed but the `CxPlatZeroMemory` right below is still there. Same for multiple other occurrences in the file.

Comment (guhetier) [reply src/core/unittest/CubicTest.cpp]: On top of this, there are generally multiple fields that are always set to a default value below.
According to Copilot review, 500 lines (~18%) of the file is redundant initialization (I didn't check the numbers :)), it seems like an area to improve.

File: src/core/unittest/CubicTest.cpp [hunk 9]
```diff
@@ -371,7 +405,9 @@ TEST(CubicTest, GetterFunctions)
     // Test GetCongestionWindow
     uint32_t CongestionWindow = Connection.CongestionControl.QuicCongestionControlGetCongestionWindow(&Connection.CongestionControl);
     ASSERT_EQ(CongestionWindow, Cubic->CongestionWindow);
```
Comment (guhetier) [inline src/core/unittest/CubicTest.cpp]: nit: Re-reading through the code, this assertion is now IMO needlessly intrusive.

What matters is that `QuicCongestionControlGetCongestionWindow` returns the expected value.
How it does it internally doesn't matter, it could be a trivial getter or make some computation.

This is a general pattern, I don't know if that's something DeepTest can get smarter about.
The tests should ideally not depend on the internal implementation details more than needed: this allows us to change the code without changing the test and get confirmation that the behavior didn't regress. If we need to modify both at once, we lose some validation.

File: src/core/unittest/CubicTest.cpp [hunk 10]
```diff
@@ -537,8 +598,10 @@ TEST(CubicTest, OnDataAcknowledged_BasicAck)
     Connection.CongestionControl.QuicCongestionControlOnDataAcknowledged(
         &Connection.CongestionControl,
         &AckEvent);
-    // Verify window may have grown (depends on slow start vs congestion avoidance)
-    ASSERT_GE(Cubic->CongestionWindow, InitialWindow);
+
+    // Window unchanged: LargestAck (5) >= HyStartRoundEnd (0) triggers round boundary
+    // crossing which resets HyStartAckCount to 0. With insufficient samples, no growth.
+    ASSERT_EQ(Cubic->CongestionWindow, InitialWindow);
```
Comment (guhetier) [inline src/core/unittest/CubicTest.cpp]: **[AI-generated]** This test passes for the wrong reason. The comment on line 602 claims *"HyStart round boundary crossing resets HyStartAckCount, preventing window growth"*, but HyStart is **disabled** in this test (`InitializeMockConnection` sets `HyStartEnabled = FALSE` and this test never overrides it). The HyStart code path at `cubic.c:476` is skipped entirely.

What actually happens: the window **does** grow from 12320 to 17320 during slow start (`+= BytesAcked`), but is then **clamped back** to `2 * BytesInFlightMax = 2 * 6160 = 12320` at `cubic.c:683`. The clamping fires because `BytesInFlight` is set directly (line 578: `Cubic->BytesInFlight = 5000`) instead of via `OnDataSent`, so `BytesInFlightMax` is never updated from its init value (`CongestionWindow / 2 = 6160`).

This test accidentally validates the `BytesInFlightMax` clamping guard, not HyStart behavior. If someone modifies the clamping logic, this test will fail for a reason unrelated to its stated purpose.

**Suggestion:** Either use `OnDataSent` to set BytesInFlight (which updates BytesInFlightMax), fix the comment to describe the real behavior, or restructure the test to actually test what it claims.

File: src/core/unittest/CubicTest.cpp [hunk 11]
```diff
@@ -759,13 +840,2412 @@ TEST(CubicTest, HyStart_StateTransitions)
...
+    AckEvent.AdjustedAckTime = AckEvent.TimeNow;
+    AckEvent.AckedPackets = NULL;
+
+    Connection.CongestionControl.QuicCongestionControlOnDataAcknowledged(
+        &Connection.CongestionControl,
+        &AckEvent);
+
+    // Accumulator is 0: LargestAck (15) > RecoverySentPacketNumber (10) exits recovery,
+    // but code jumps to Exit label after clearing IsInRecovery, so AIMD logic not reached.
+    ASSERT_EQ(Cubic->AimdAccumulator, 0u);
```
Comment (guhetier) [inline src/core/unittest/CubicTest.cpp]: **[AI-generated]** Tests 24 and 25 (`AIMD_AccumulatorBelowWindowPrior` and `AIMD_AccumulatorAboveWindowPrior`) both claim to test AIMD accumulator logic (0.5 vs 1.0 MSS/RTT slope), but neither test actually exercises that code.

In both tests, `LargestAck=15 > RecoverySentPacketNumber=10`, so recovery exits at `cubic.c:454`, which then `goto Exit` at line 468 **before** reaching the AIMD accumulator logic at `cubic.c:649`. Both tests only verify `AimdAccumulator == 0` after recovery exit — they are functionally identical (differing only in BytesAcked: 600 vs 1200, which doesn't matter since the AIMD code is never reached).

The AIMD accumulator paths (`AimdAccumulator += BytesAcked / 2` when `AimdWindow < WindowPrior`, vs `AimdAccumulator += BytesAcked` otherwise) have **no test coverage**.

**Suggestion:** To actually test AIMD, exit recovery first (via a prior ACK with `LargestAck > RecoverySentPacketNumber`), then send a subsequent ACK with non-zero `BytesAcked` that will reach the congestion avoidance code at `cubic.c:563+`.

File: src/core/unittest/CubicTest.cpp [hunk 12]
```diff
@@ -759,13 +840,2412 @@ TEST(CubicTest, HyStart_StateTransitions)
...
 
-    // HyStart may transition states based on RTT measurements
-    // Just verify state is valid and divisor is set appropriately
-    ASSERT_TRUE(Cubic->HyStartState >= HYSTART_NOT_STARTED &&
-                Cubic->HyStartState <= HYSTART_DONE);
-    ASSERT_GE(Cubic->CWndSlowStartGrowthDivisor, 1u);
+    // LargestAck (10) is NOT > RecoverySentPacketNumber (10), so we stay in recovery
+    // Window doesn't grow during recovery - just BytesInFlight decreases
+    ASSERT_TRUE(Cubic->IsInRecovery);
+    ASSERT_EQ(Cubic->CongestionWindow, WindowBefore);
```
Comment (guhetier) [inline src/core/unittest/CubicTest.cpp]: **[AI-generated]** This test claims to test *"AIMD vs CUBIC window selection logic"* but it stays in recovery the entire time. `LargestAck=10` is NOT `> RecoverySentPacketNumber=10` (the comparison is strict `>`), so recovery is never exited and the code jumps to `Exit` at `cubic.c:468`.

The AIMD vs CUBIC selection logic at `cubic.c:659-670` (`if (Cubic->AimdWindow > CubicWindow)`) is never reached. The only thing this test validates is that the window doesn't change during recovery — which is already covered by tests 26 and 30.

**Suggestion:** Exit recovery first, then ACK data to exercise the actual AIMD vs CUBIC window selection in congestion avoidance. Or rename this test to accurately describe what it tests (e.g., `Recovery_WindowUnchangedDuringRecovery`).

File: src/core/unittest/CubicTest.cpp [hunk 13]
```diff
@@ -759,13 +840,2412 @@ TEST(CubicTest, HyStart_StateTransitions)
...
+    AckEvent.AckedPackets = NULL;
+
+    Connection.CongestionControl.QuicCongestionControlOnDataAcknowledged(
+        &Connection.CongestionControl,
+        &AckEvent);
+
+    // Still in recovery because LargestAck (10) is NOT > RecoverySentPacketNumber (10)
+    // Window stays the same during recovery
+    ASSERT_TRUE(Cubic->IsInRecovery);
+    ASSERT_EQ(Cubic->CongestionWindow, WindowAfterLoss);
```
Comment (guhetier) [inline src/core/unittest/CubicTest.cpp]: **[AI-generated]** This test claims to test *"overflow handling in CUBIC window calculation"* but it stays in recovery (`LargestAck=10` NOT `> RecoverySentPacketNumber=10`). The CUBIC window formula at `cubic.c:620` and the overflow check at `cubic.c:625` (`if (CubicWindow < 0)`) are never reached.

The test only verifies that the window doesn't change during recovery — identical to what tests 23 and 30 already verify.

**Suggestion:** Exit recovery before the ACK that should trigger the CUBIC window overflow, so the congestion avoidance code path is actually exercised.

File: src/core/unittest/CubicTest.cpp [hunk 14]
```diff
@@ -759,13 +840,2412 @@ TEST(CubicTest, HyStart_StateTransitions)
...
+    AckEvent2.AckedPackets = NULL;
+
+    Connection.CongestionControl.QuicCongestionControlOnDataAcknowledged(
+        &Connection.CongestionControl,
+        &AckEvent2);
+
+    // Still in recovery because LargestAck (10) is NOT > RecoverySentPacketNumber (10)
+    // Window stays the same during recovery
+    ASSERT_TRUE(Cubic->IsInRecovery);
+    ASSERT_EQ(Cubic->CongestionWindow, WindowAfterFirst);
```
Comment (guhetier) [inline src/core/unittest/CubicTest.cpp]: **[AI-generated]** This test claims to test *"behavior with large time gaps between ACKs"* (idle period handling), but both ACKs stay in recovery (`LargestAck` is never `> RecoverySentPacketNumber=10`). The time gap adjustment logic at `cubic.c:580-588` is only reached **outside** recovery (the `IsInRecovery` check at `cubic.c:453` causes a `goto Exit` before the time gap code).

The 5-second gap between ACK1 and ACK2 has no effect — the test just verifies the window doesn't change during recovery, same as tests 23 and 26.

**Suggestion:** Exit recovery, establish `TimeOfLastAckValid=TRUE` with an initial ACK, then send a second ACK after a long idle gap to actually exercise the `TimeSinceLastAck > SendIdleTimeoutMs` path at `cubic.c:582`.

**Review by guhetier — COMMENTED** [review]
AI-generated review (GitHub Copilot CLI). Focused on test correctness: 6 high-severity issues where tests don't exercise the code paths they claim to test.

File: src/core/unittest/CubicTest.cpp [hunk 15]
```diff
@@ -759,13 +840,2412 @@ TEST(CubicTest, HyStart_StateTransitions)
...
+    LossEvent.NumRetransmittableBytes = 2400;
+    LossEvent.PersistentCongestion = FALSE;
+    LossEvent.LargestPacketNumberLost = 5;
+    LossEvent.LargestSentPacketNumber = 10;
+
+    Connection.CongestionControl.QuicCongestionControlOnDataLost(&Connection.CongestionControl, &LossEvent);
+
+    // Invariant 2: DONE → divisor = 1
+    ASSERT_EQ(Cubic->HyStartState, HYSTART_DONE);
+    ASSERT_EQ(Cubic->CWndSlowStartGrowthDivisor, 1u);
```
Comment (guhetier) [inline src/core/unittest/CubicTest.cpp]: **[AI-generated]** This test claims to verify the growth divisor invariant across all three HyStart states (*"NOT_STARTED → divisor=1, ACTIVE → divisor=4, DONE → divisor=1"*), but it only transitions NOT_STARTED → DONE via loss. The ACTIVE state (divisor=4) is **never entered or verified** in this test.

Test 42 (`HyStart_DelayIncreaseDetection_TriggerActiveTransition`) already verifies the ACTIVE state divisor at line 2669.

**Suggestion:** Either add the ACTIVE state transition to this test (transition NOT_STARTED → ACTIVE → DONE and check divisor at each step), or update the test description to reflect that it only verifies NOT_STARTED and DONE states.

File: src/core/unittest/CubicTest.cpp (hunk 8 above)
Comment (saikat107) [reply src/core/unittest/CubicTest.cpp]: Done in the new version.

File: src/core/unittest/CubicTest.cpp (hunk 9 above)
Comment (saikat107) [reply src/core/unittest/CubicTest.cpp]: In the current version, expected values were calculated based on formulae, and used APIs to manistates. Hope this is better.

File: src/core/unittest/CubicTest.cpp [hunk 16]
```diff
@@ -23,749 +25,2394 @@ static void InitializeMockConnection(
...
-    // Set to max
-    Connection.CongestionControl.QuicCongestionControlSetExemption(&Connection.CongestionControl, 255);
-    ASSERT_EQ(Cubic->Exemptions, 255u);
-}
+// //
+// // Test: SetExemption
+// // Scenario: Tests SetExemption to verify it correctly sets the number of packets that
+// // can bypass congestion control. Used for probe packets and other special cases.
+// //
+// TEST_F(CubicTest, SetExemption)
```
Comment (guhetier) [inline src/core/unittest/CubicTest.cpp]: Commented test?

File: src/core/unittest/CubicTest.cpp [hunk 17]
```diff
@@ -23,749 +25,2394 @@ static void InitializeMockConnection(
...
+    CC->QuicCongestionControlOnDataAcknowledged(CC, &ExitAck);
+    ASSERT_FALSE(Cubic->IsInRecovery);
+
+    // Force extreme conditions to trigger CUBIC overflow.
+    // Set a huge WindowMax so KCubic is very large, making DeltaT deeply negative
+    // when TimeInCongAvoid is small. The cubic term DeltaT^3 overflows negative.
+    Cubic->WindowMax = UINT32_MAX;
+    // Recalculate KCubic for the extreme WindowMax (mirrors the loss handler logic)
+    Cubic->KCubic =
+        CubeRoot(
```
Comment (guhetier) [inline src/core/unittest/CubicTest.cpp]: It seems weird to me to pull a private function from the cubic compilation unit just to set a test value.
Pulling a function in to test it would have been a thing, but here, it is used as a helper to set test data, and combined with a "magic formula".

At this point, why not set a value directly in KCubic, that will trigger the scenario we care about and explain why this value? (like it is done for WindowsMax)?

File: src/core/unittest/CubicTest.cpp (hunk 16 above)
Comment (saikat107) [reply src/core/unittest/CubicTest.cpp]: Mistake

**Comment by saikat107** [comment]
Check this out, @guhetier

## Overview

| Category | Count | % of 50 |
|----------|-------|---------|
| Specification-conformance | 23 | 46% |
| Implementation-specific, loosely coupled | 7 | 14% |
| Implementation-specific, tightly coupled | 20 | 40% |

**Specification-conformance** tests validate behavior defined by RFC 8312 (CUBIC), RFC 9002 (QUIC Loss Detection), or the HyStart++ draft. These tests would remain valid across any conforming implementation.

**Implementation-specific, loosely coupled** tests exercise MsQuic-specific features (pacing API, exemptions, statistics) but assert only through public API outputs. They would survive an internal refactor.

**Implementation-specific, tightly coupled** tests directly read or write internal `Cubic->` struct fields. If the implementation is refactored — even while preserving identical congestion control behavior — these tests would break.

---

## Specification-Conformance Tests (23)

These test behavior mandated or described by protocol specifications.

| Test | Spec Basis |
|------|-----------|
| `OnDataLost_WindowReduction` | RFC 8312 §4.6: β_cubic = 0.7 multiplicative decrease |
| `OnEcn_CongestionSignal` | RFC 9002 §7.1: ECN treated as congestion signal |
| `FastConvergence_AdditionalReduction` | RFC 8312 §4.7: fast convergence algorithm |
| `Recovery_ExitOnNewAck` | RFC 9002 §7.3.2: exit recovery on ACK of new data |
| `OnDataAcknowledged_BasicAck` | RFC 8312 §4.1: slow start window growth |
| `CongestionAvoidance_AIMDvsCubicSelection` | RFC 8312 §4.2–4.4: W_cubic vs W_est selection |
| `AIMD_SequentialLinearConvergence` | RFC 8312 §4.3: TCP-friendly estimation |
| `CongestionAvoidance_RenoFriendlyRegion` | RFC 8312 §4.3: Reno-friendly region |
| `CongestionAvoidance_BoundedGrowthClamp` | RFC 8312 §4.2: bounded window growth |
| `MinimumWindowFloor_AfterLoss` | RFC 9002 §7.2: minimum congestion window of 2×MTU |
| `SlowStart_BytesInFlightMaxClamp` | RFC 9002 §7.8: pacing/limiting based on bytes in flight |
| `Recovery_DoubleLossNoDoubleReduction` | RFC 9002 §7.3.1: single reduction per recovery period |
| `Recovery_StayOnOldAck` | RFC 9002 §7.3.2: stay in recovery until new ACK |
| `SpuriousCongestion_StateRollback` | RFC 8312 §4.8: response to spurious congestion |
| `HyStart_NotStartedToDone_ViaLoss` | HyStart++ draft: loss triggers transition to DONE |
| `HyStart_NotStartedToDone_ViaECN` | HyStart++ draft: ECN triggers transition to DONE |
| `HyStart_ActiveToDone_ViaLoss` | HyStart++ draft: loss from ACTIVE triggers DONE |
| `HyStart_ActiveToDone_ViaECN` | HyStart++ draft: ECN from ACTIVE triggers DONE |
| `HyStart_AnyToDone_ViaPersistentCongestion` | HyStart++ draft: persistent congestion triggers DONE |
| `HyStart_TerminalState_DoneIsAbsorbing` | HyStart++ draft: DONE is a terminal state |
| `HyStart_DelayIncreaseDetection_TriggerActiveTransition` | HyStart++ draft: delay increase → ACTIVE |
| `HyStart_RttDecreaseDetection_ReturnToNotStarted` | HyStart++ draft: RTT decrease → NOT_STARTED |
| `HyStart_ConservativeSlowStartRounds_TransitionToDone` | HyStart++ draft: conservative rounds → DONE |

---

## Implementation-Specific Tests — Loosely Coupled (7)

These test MsQuic-specific features but assert only through public API outputs (`CanSend`, `GetSendAllowance`, `GetNetworkStatistics`, etc.). They would survive an internal refactor that preserves the same API contract.

| Test | Asserts On |
|------|-----------|
| `CanSendScenarios` | `CanSend()` return value |
| `GetSendAllowanceScenarios` | `GetSendAllowance()` return value |
| `CongestionWindowCalculation` | `CongestionWindow` (primary output) |
| `GetNetworkStatistics_RetrieveStats` | `GetNetworkStatistics()` output struct |
| `CongestionControlOnSpuriousCongestionEvent` | `OnSpuriousCongestionEvent()` return value |
| `ZeroBytesAcked_EarlyExit` | `CongestionWindow` unchanged after zero-byte ACK |
| `AppLimited_APICoverage` | `IsAppLimited()` / `SetAppLimited()` return values |

---

## Implementation-Specific Tests — Tightly Coupled (20)

These tests directly read or write internal `QUIC_CONGESTION_CONTROL_CUBIC` struct fields (e.g., `AimdAccumulator`, `TimeOfCongAvoidStart`, `HyStartState`, `PrevCongestionWindow`). If the implementation is refactored while preserving identical congestion control behavior, these tests would fail.

| Test | Internal Fields Touched |
|------|------------------------|
| `ResetScenarios` | `MinRttInLastRound` (asserts sentinel value) |
| `OnDataSent_IncrementsBytesInFlight` | `BytesInFlight`, `BytesInFlightMax`, `NumRetransmittableExemptions` |
| `OnDataInvalidated_DecrementsBytesInFlight` | `BytesInFlight` |
| `Pacing_SlowStartWindowEstimation` | Writes `SlowStartThreshold`; derives expected from exact `EstWnd = 2×CW` formula |
| `Pacing_CongestionAvoidanceEstimation` | Derives expected from exact `EstWnd = 1.25×CW` formula |
| `Pacing_LastSendAllowanceCarryover` | Tests `LastSendAllowance` accumulation mechanism |
| `AIMD_AccumulatorBelowWindowPrior` | Asserts `AimdAccumulator` exact value (half-rate path) |
| `AIMD_AccumulatorAboveWindowPrior` | Asserts `AimdAccumulator` exact value (full-rate path) |
| `AIMD_AccumulatorTriggersWindowGrowth` | Asserts `AimdAccumulator` overflow triggers +DPL |
| `CubicWindow_OverflowToBytesInFlightMax` | Writes `WindowMax`, `KCubic`, `TimeOfLastAckValid` |
| `TimeGap_IdlePeriodHandling` | Asserts `TimeOfCongAvoidStart` adjustment |
| `CongestionAvoidance_TimeGapOverflowProtection` | Writes and asserts `TimeOfCongAvoidStart` |
| `ECN_SpuriousRollbackRestoresStaleState` | Asserts `PrevCongestionWindow`, `PrevAimdWindow`, `PrevSlowStartThreshold` |
| `SlowStart_WindowOverflowAfterPersistentCongestion` | Asserts `TimeOfCongAvoidStart`, `AimdWindow` |
| `HyStart_InitialStateVerification` | `HyStartState`, `CWndSlowStartGrowthDivisor`, `HyStartAckCount`, etc. |
| `HyStart_Disabled_NoTransitions` | `HyStartState` |
| `HyStart_StateInvariant_GrowthDivisor` | `HyStartState`, `CWndSlowStartGrowthDivisor` |
| `HyStart_RecoveryExit_StatePersistence` | `HyStartState` |
| `HyStart_SpuriousCongestion_StateNotRolledBack` | `HyStartState` |
| `HyStart_DelayIncreaseDetection_EtaCalculationAndCondition` | `HyStartAckCount`, `MinRttInCurrentRound`, `HyStartState` |

---

## CubeRootTest (separate test suite)

16 parametric test cases (`CubeRootCases/CubeRootTest.ReturnsExpectedValue/0..15`) test the `CubeRoot` utility function with boundary values, perfect cubes, non-perfect cubes, and large inputs. These are **specification-conformance** tests — they validate a pure mathematical function against known correct values.
---

Shall we remove the test of the last class of tests?

**Comment by guhetier** [comment]
@saikat107 Nice categorization. I think all categories can stay in, ideally we can separate them with a banner comment:

```
//
//  Implementation specific test - tightly coupled
//
```

so future devs know what to expect

**Comment by guhetier** [comment]
@saikat107 Can you please rebase the PR one last time? We fixed the cargo build, but this PR will need to pickup the change to complete the CI.
All looks good otherwise.
